LinkList.java

```diff
@@ -120,34 +133,153 @@ public boolean contains(Object o) {
 
      }
 
-     public static void main(String [ ] args) {
-        ArrayList a=new LinkList();
+      public Iterator<E> iterator() {
+        return null;
+    }  
+
+    public void addFirst(E e) {
+        if(size()==0){
+            head.data=e;
+            numNodes++;
+            return;
+        }
+        Node holder=new Node<E>();
+        holder.data=head.data;
+        holder.next=head.next;
+        head.data=e;
+        head.next=holder;
+        numNodes++;
+    }
+
+    public void addLast(E e) {
+        if(size()==0){
+            head.data=e;
+            numNodes++;
+            return;
+        }
+        Node tra=head;
+        while(tra.next!=null) {
+            tra=tra.next;
+        }
+        Node temp=new Node<E>();
+        temp.data=e;
+        tra.next=temp;
+        numNodes++;
+    }
+
+    @SuppressWarnings("unchecked")
+    public E getFirst() {
+        if(size()==0){
+            return null;
+        }
+        return (E)head.data;
+    }
+
+    @SuppressWarnings("unchecked")
+    public E getLast() {
+        if(size()==0){
+            return null;
+        }
+        Node temp=head;
+        while(temp.next!=null) {
+            temp=temp.next;
+        }
+        return (E)temp.data;
+    }
+
+    @SuppressWarnings("unchecked")
+    public E removeFirst() {
+        if(size()==0){
+            throw new NoSuchElementException();
+        }
+        E holder=(E)head.data;
+        head=head.next;
+        numNodes--;
+        return holder;
+        
+    }
+
+    public E removeLast() {
+        if(size()==0){
+            throw new NoSuchElementException();
+        }
+        Node f=head.next;
+        Node r=head;
+        while(f.next!=null) {
+            f=f.next;
+            r=r.next;
+        }
+        r.next=null;
+        numNodes--;
+        return (E)r.data;
+    }
+
+    public boolean offer(E e) {//add element as the tail of the list
+        if(size()==0){
+            head.data=e;
+            numNodes++;
+            return true;
+        }
+        addLast(e);
+        return true;
+    }
+
+    public E element() {//retrives the head of the list
+        return getFirst();
+    }
+
+
+    public E peek() {
+        if(isEmpty()) {
```

Comment:
Good. This is better than if(size()==0) as the semantic is more clear

Reply:
ok